CI: Data Race in TestManager

See https://circleci.com/gh/docker/swarm-v2/1386

```
==================
WARNING: DATA RACE
Write by goroutine 21:
  runtime.mapassign1()
      /usr/local/go/src/runtime/hashmap.go:429 +0x0
  github.com/docker/swarm-v2/vendor/google.golang.org/grpc.(*Server).register()
      /home/ubuntu/.go_workspace/src/github.com/docker/swarm-v2/vendor/google.golang.org/grpc/server.go:215 +0x6fc
  github.com/docker/swarm-v2/vendor/google.golang.org/grpc.(*Server).RegisterService()
      /home/ubuntu/.go_workspace/src/github.com/docker/swarm-v2/vendor/google.golang.org/grpc/server.go:192 +0x3da
  github.com/docker/swarm-v2/api.RegisterManagerServer()
      /home/ubuntu/.go_workspace/src/github.com/docker/swarm-v2/api/manager.pb.go:420 +0x6a
  github.com/docker/swarm-v2/manager.(*Manager).Run()
      github.com/docker/swarm-v2/manager/_test/_obj_test/manager.go:283 +0x14c4
  github.com/docker/swarm-v2/manager.TestManager.func1()
      /home/ubuntu/.go_workspace/src/github.com/docker/swarm-v2/manager/manager_test.go:66 +0x42

Previous read by goroutine 32:
  runtime.mapaccess2_faststr()
      /usr/local/go/src/runtime/hashmap_fast.go:298 +0x0
  github.com/docker/swarm-v2/vendor/google.golang.org/grpc.(*Server).handleStream()
      /home/ubuntu/.go_workspace/src/github.com/docker/swarm-v2/vendor/google.golang.org/grpc/server.go:626 +0xa7f
  github.com/docker/swarm-v2/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1()
      /home/ubuntu/.go_workspace/src/github.com/docker/swarm-v2/vendor/google.golang.org/grpc/server.go:323 +0xad

Goroutine 21 (running) created at:
  github.com/docker/swarm-v2/manager.TestManager()
      /home/ubuntu/.go_workspace/src/github.com/docker/swarm-v2/manager/manager_test.go:67 +0x937
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:473 +0xdc

Goroutine 32 (running) created at:
  github.com/docker/swarm-v2/vendor/google.golang.org/grpc.(*Server).serveStreams.func1()
      /home/ubuntu/.go_workspace/src/github.com/docker/swarm-v2/vendor/google.golang.org/grpc/server.go:324 +0xa7
  github.com/docker/swarm-v2/vendor/google.golang.org/grpc/transport.(*http2Server).operateHeaders()
      /home/ubuntu/.go_workspace/src/github.com/docker/swarm-v2/vendor/google.golang.org/grpc/transport/http2_server.go:213 +0x19c8
  github.com/docker/swarm-v2/vendor/google.golang.org/grpc/transport.(*http2Server).HandleStreams()
      /home/ubuntu/.go_workspace/src/github.com/docker/swarm-v2/vendor/google.golang.org/grpc/transport/http2_server.go:272 +0xf56
  github.com/docker/swarm-v2/vendor/google.golang.org/grpc.(*Server).serveStreams()
      /home/ubuntu/.go_workspace/src/github.com/docker/swarm-v2/vendor/google.golang.org/grpc/server.go:325 +0x1dc
  github.com/docker/swarm-v2/vendor/google.golang.org/grpc.(*Server).serveNewHTTP2Transport()
      /home/ubuntu/.go_workspace/src/github.com/docker/swarm-v2/vendor/google.golang.org/grpc/server.go:312 +0x54d
  github.com/docker/swarm-v2/vendor/google.golang.org/grpc.(*Server).handleRawConn()
      /home/ubuntu/.go_workspace/src/github.com/docker/swarm-v2/vendor/google.golang.org/grpc/server.go:289 +0x5b1
==================
--- FAIL: TestManager (0.09s)
    Error Trace:    manager_test.go:84
    Error:      Not equal: "unknown service docker.cluster.api.Dispatcher" (expected)
                    != "node not registered" (actual)
```

--- a/api/ca.pb.go
+++ b/api/ca.pb.go
@@ -523,16 +523,16 @@ func encodeVarintCa(data []byte, offset int, v uint64) int {
 }
 
 type raftProxyCAServer struct {
-	local   CAServer
-	conn    *grpc.ClientConn
-	cluster raftpicker.RaftCluster
+	local        CAServer
+	connSelector *raftpicker.ConnSelector
+	cluster      raftpicker.RaftCluster
 }
 
-func NewRaftProxyCAServer(local CAServer, conn *grpc.ClientConn, cluster raftpicker.RaftCluster) CAServer {
+func NewRaftProxyCAServer(local CAServer, connSelector *raftpicker.ConnSelector, cluster raftpicker.RaftCluster) CAServer {
 	return &raftProxyCAServer{
-		local:   local,
-		cluster: cluster,
-		conn:    conn,
+		local:        local,
+		cluster:      cluster,
+		connSelector: connSelector,
 	}
 }
 
@@ -556,7 +556,11 @@ func (p *raftProxyCAServer) IssueCertificate(ctx context.Context, r *IssueCertif
 	md["redirect"] = append(md["redirect"], addr)
 	ctx = metadata.NewContext(ctx, md)
 
-	return NewCAClient(p.conn).IssueCertificate(ctx, r)
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return nil, err
+	}
+	return NewCAClient(conn).IssueCertificate(ctx, r)
 }
 
 func (p *raftProxyCAServer) GetRootCACertificate(ctx context.Context, r *GetRootCACertificateRequest) (*GetRootCACertificateResponse, error) {
@@ -579,7 +583,11 @@ func (p *raftProxyCAServer) GetRootCACertificate(ctx context.Context, r *GetRoot
 	md["redirect"] = append(md["redirect"], addr)
 	ctx = metadata.NewContext(ctx, md)
 
-	return NewCAClient(p.conn).GetRootCACertificate(ctx, r)
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return nil, err
+	}
+	return NewCAClient(conn).GetRootCACertificate(ctx, r)
 }
 
 func (m *IssuanceStatus) Size() (n int) {

--- a/api/cluster.pb.go
+++ b/api/cluster.pb.go
@@ -3121,16 +3121,16 @@ func encodeVarintCluster(data []byte, offset int, v uint64) int {
 }
 
 type raftProxyClusterServer struct {
-	local   ClusterServer
-	conn    *grpc.ClientConn
-	cluster raftpicker.RaftCluster
+	local        ClusterServer
+	connSelector *raftpicker.ConnSelector
+	cluster      raftpicker.RaftCluster
 }
 
-func NewRaftProxyClusterServer(local ClusterServer, conn *grpc.ClientConn, cluster raftpicker.RaftCluster) ClusterServer {
+func NewRaftProxyClusterServer(local ClusterServer, connSelector *raftpicker.ConnSelector, cluster raftpicker.RaftCluster) ClusterServer {
 	return &raftProxyClusterServer{
-		local:   local,
-		cluster: cluster,
-		conn:    conn,
+		local:        local,
+		cluster:      cluster,
+		connSelector: connSelector,
 	}
 }
 
@@ -3154,7 +3154,11 @@ func (p *raftProxyClusterServer) GetNode(ctx context.Context, r *GetNodeRequest)
 	md["redirect"] = append(md["redirect"], addr)
 	ctx = metadata.NewContext(ctx, md)
 
-	return NewClusterClient(p.conn).GetNode(ctx, r)
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return nil, err
+	}
+	return NewClusterClient(conn).GetNode(ctx, r)
 }
 
 func (p *raftProxyClusterServer) ListNodes(ctx context.Context, r *ListNodesRequest) (*ListNodesResponse, error) {
@@ -3177,7 +3181,11 @@ func (p *raftProxyClusterServer) ListNodes(ctx context.Context, r *ListNodesRequ
 	md["redirect"] = append(md["redirect"], addr)
 	ctx = metadata.NewContext(ctx, md)
 
-	return NewClusterClient(p.conn).ListNodes(ctx, r)
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return nil, err
+	}
+	return NewClusterClient(conn).ListNodes(ctx, r)
 }
 
 func (p *raftProxyClusterServer) UpdateNode(ctx context.Context, r *UpdateNodeRequest) (*UpdateNodeResponse, error) {
@@ -3200,7 +3208,11 @@ func (p *raftProxyClusterServer) UpdateNode(ctx context.Context, r *UpdateNodeRe
 	md["redirect"] = append(md["redirect"], addr)
 	ctx = metadata.NewContext(ctx, md)
 
-	return NewClusterClient(p.conn).UpdateNode(ctx, r)
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return nil, err
+	}
+	return NewClusterClient(conn).UpdateNode(ctx, r)
 }
 
 func (p *raftProxyClusterServer) GetTask(ctx context.Context, r *GetTaskRequest) (*GetTaskResponse, error) {
@@ -3223,7 +3235,11 @@ func (p *raftProxyClusterServer) GetTask(ctx context.Context, r *GetTaskRequest)
 	md["redirect"] = append(md["redirect"], addr)
 	ctx = metadata.NewContext(ctx, md)
 
-	return NewClusterClient(p.conn).GetTask(ctx, r)
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return nil, err
+	}
+	return NewClusterClient(conn).GetTask(ctx, r)
 }
 
 func (p *raftProxyClusterServer) ListTasks(ctx context.Context, r *ListTasksRequest) (*ListTasksResponse, error) {
@@ -3246,7 +3262,11 @@ func (p *raftProxyClusterServer) ListTasks(ctx context.Context, r *ListTasksRequ
 	md["redirect"] = append(md["redirect"], addr)
 	ctx = metadata.NewContext(ctx, md)
 
-	return NewClusterClient(p.conn).ListTasks(ctx, r)
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return nil, err
+	}
+	return NewClusterClient(conn).ListTasks(ctx, r)
 }
 
 func (p *raftProxyClusterServer) RemoveTask(ctx context.Context, r *RemoveTaskRequest) (*RemoveTaskResponse, error) {
@@ -3269,7 +3289,11 @@ func (p *raftProxyClusterServer) RemoveTask(ctx context.Context, r *RemoveTaskRe
 	md["redirect"] = append(md["redirect"], addr)
 	ctx = metadata.NewContext(ctx, md)
 
-	return NewClusterClient(p.conn).RemoveTask(ctx, r)
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return nil, err
+	}
+	return NewClusterClient(conn).RemoveTask(ctx, r)
 }
 
 func (p *raftProxyClusterServer) GetService(ctx context.Context, r *GetServiceRequest) (*GetServiceResponse, error) {
@@ -3292,7 +3316,11 @@ func (p *raftProxyClusterServer) GetService(ctx context.Context, r *GetServiceRe
 	md["redirect"] = append(md["redirect"], addr)
 	ctx = metadata.NewContext(ctx, md)
 
-	return NewClusterClient(p.conn).GetService(ctx, r)
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return nil, err
+	}
+	return NewClusterClient(conn).GetService(ctx, r)
 }
 
 func (p *raftProxyClusterServer) ListServices(ctx context.Context, r *ListServicesRequest) (*ListServicesResponse, error) {
@@ -3315,7 +3343,11 @@ func (p *raftProxyClusterServer) ListServices(ctx context.Context, r *ListServic
 	md["redirect"] = append(md["redirect"], addr)
 	ctx = metadata.NewContext(ctx, md)
 
-	return NewClusterClient(p.conn).ListServices(ctx, r)
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return nil, err
+	}
+	return NewClusterClient(conn).ListServices(ctx, r)
 }
 
 func (p *raftProxyClusterServer) CreateService(ctx context.Context, r *CreateServiceRequest) (*CreateServiceResponse, error) {
@@ -3338,7 +3370,11 @@ func (p *raftProxyClusterServer) CreateService(ctx context.Context, r *CreateSer
 	md["redirect"] = append(md["redirect"], addr)
 	ctx = metadata.NewContext(ctx, md)
 
-	return NewClusterClient(p.conn).CreateService(ctx, r)
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return nil, err
+	}
+	return NewClusterClient(conn).CreateService(ctx, r)
 }
 
 func (p *raftProxyClusterServer) UpdateService(ctx context.Context, r *UpdateServiceRequest) (*UpdateServiceResponse, error) {
@@ -3361,7 +3397,11 @@ func (p *raftProxyClusterServer) UpdateService(ctx context.Context, r *UpdateSer
 	md["redirect"] = append(md["redirect"], addr)
 	ctx = metadata.NewContext(ctx, md)
 
-	return NewClusterClient(p.conn).UpdateService(ctx, r)
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return nil, err
+	}
+	return NewClusterClient(conn).UpdateService(ctx, r)
 }
 
 func (p *raftProxyClusterServer) RemoveService(ctx context.Context, r *RemoveServiceRequest) (*RemoveServiceResponse, error) {
@@ -3384,7 +3424,11 @@ func (p *raftProxyClusterServer) RemoveService(ctx context.Context, r *RemoveSer
 	md["redirect"] = append(md["redirect"], addr)
 	ctx = metadata.NewContext(ctx, md)
 
-	return NewClusterClient(p.conn).RemoveService(ctx, r)
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return nil, err
+	}
+	return NewClusterClient(conn).RemoveService(ctx, r)
 }
 
 func (p *raftProxyClusterServer) GetNetwork(ctx context.Context, r *GetNetworkRequest) (*GetNetworkResponse, error) {
@@ -3407,7 +3451,11 @@ func (p *raftProxyClusterServer) GetNetwork(ctx context.Context, r *GetNetworkRe
 	md["redirect"] = append(md["redirect"], addr)
 	ctx = metadata.NewContext(ctx, md)
 
-	return NewClusterClient(p.conn).GetNetwork(ctx, r)
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return nil, err
+	}
+	return NewClusterClient(conn).GetNetwork(ctx, r)
 }
 
 func (p *raftProxyClusterServer) ListNetworks(ctx context.Context, r *ListNetworksRequest) (*ListNetworksResponse, error) {
@@ -3430,7 +3478,11 @@ func (p *raftProxyClusterServer) ListNetworks(ctx context.Context, r *ListNetwor
 	md["redirect"] = append(md["redirect"], addr)
 	ctx = metadata.NewContext(ctx, md)
 
-	return NewClusterClient(p.conn).ListNetworks(ctx, r)
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return nil, err
+	}
+	return NewClusterClient(conn).ListNetworks(ctx, r)
 }
 
 func (p *raftProxyClusterServer) CreateNetwork(ctx context.Context, r *CreateNetworkRequest) (*CreateNetworkResponse, error) {
@@ -3453,7 +3505,11 @@ func (p *raftProxyClusterServer) CreateNetwork(ctx context.Context, r *CreateNet
 	md["redirect"] = append(md["redirect"], addr)
 	ctx = metadata.NewContext(ctx, md)
 
-	return NewClusterClient(p.conn).CreateNetwork(ctx, r)
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return nil, err
+	}
+	return NewClusterClient(conn).CreateNetwork(ctx, r)
 }
 
 func (p *raftProxyClusterServer) RemoveNetwork(ctx context.Context, r *RemoveNetworkRequest) (*RemoveNetworkResponse, error) {
@@ -3476,7 +3532,11 @@ func (p *raftProxyClusterServer) RemoveNetwork(ctx context.Context, r *RemoveNet
 	md["redirect"] = append(md["redirect"], addr)
 	ctx = metadata.NewContext(ctx, md)
 
-	return NewClusterClient(p.conn).RemoveNetwork(ctx, r)
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return nil, err
+	}
+	return NewClusterClient(conn).RemoveNetwork(ctx, r)
 }
 
 func (p *raftProxyClusterServer) GetVolume(ctx context.Context, r *GetVolumeRequest) (*GetVolumeResponse, error) {
@@ -3499,7 +3559,11 @@ func (p *raftProxyClusterServer) GetVolume(ctx context.Context, r *GetVolumeRequ
 	md["redirect"] = append(md["redirect"], addr)
 	ctx = metadata.NewContext(ctx, md)
 
-	return NewClusterClient(p.conn).GetVolume(ctx, r)
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return nil, err
+	}
+	return NewClusterClient(conn).GetVolume(ctx, r)
 }
 
 func (p *raftProxyClusterServer) ListVolumes(ctx context.Context, r *ListVolumesRequest) (*ListVolumesResponse, error) {
@@ -3522,7 +3586,11 @@ func (p *raftProxyClusterServer) ListVolumes(ctx context.Context, r *ListVolumes
 	md["redirect"] = append(md["redirect"], addr)
 	ctx = metadata.NewContext(ctx, md)
 
-	return NewClusterClient(p.conn).ListVolumes(ctx, r)
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return nil, err
+	}
+	return NewClusterClient(conn).ListVolumes(ctx, r)
 }
 
 func (p *raftProxyClusterServer) CreateVolume(ctx context.Context, r *CreateVolumeRequest) (*CreateVolumeResponse, error) {
@@ -3545,7 +3613,11 @@ func (p *raftProxyClusterServer) CreateVolume(ctx context.Context, r *CreateVolu
 	md["redirect"] = append(md["redirect"], addr)
 	ctx = metadata.NewContext(ctx, md)
 
-	return NewClusterClient(p.conn).CreateVolume(ctx, r)
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return nil, err
+	}
+	return NewClusterClient(conn).CreateVolume(ctx, r)
 }
 
 func (p *raftProxyClusterServer) RemoveVolume(ctx context.Context, r *RemoveVolumeRequest) (*RemoveVolumeResponse, error) {
@@ -3568,7 +3640,11 @@ func (p *raftProxyClusterServer) RemoveVolume(ctx context.Context, r *RemoveVolu
 	md["redirect"] = append(md["redirect"], addr)
 	ctx = metadata.NewContext(ctx, md)
 
-	return NewClusterClient(p.conn).RemoveVolume(ctx, r)
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return nil, err
+	}
+	return NewClusterClient(conn).RemoveVolume(ctx, r)
 }
 
 func (p *raftProxyClusterServer) ListManagers(ctx context.Context, r *ListManagersRequest) (*ListManagersResponse, error) {
@@ -3591,7 +3667,11 @@ func (p *raftProxyClusterServer) ListManagers(ctx context.Context, r *ListManage
 	md["redirect"] = append(md["redirect"], addr)
 	ctx = metadata.NewContext(ctx, md)
 
-	return NewClusterClient(p.conn).ListManagers(ctx, r)
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return nil, err
+	}
+	return NewClusterClient(conn).ListManagers(ctx, r)
 }
 
 func (m *ListOptions) Size() (n int) {

--- a/api/dispatcher.pb.go
+++ b/api/dispatcher.pb.go
@@ -1104,16 +1104,16 @@ func encodeVarintDispatcher(data []byte, offset int, v uint64) int {
 }
 
 type raftProxyDispatcherServer struct {
-	local   DispatcherServer
-	conn    *grpc.ClientConn
-	cluster raftpicker.RaftCluster
+	local        DispatcherServer
+	connSelector *raftpicker.ConnSelector
+	cluster      raftpicker.RaftCluster
 }
 
-func NewRaftProxyDispatcherServer(local DispatcherServer, conn *grpc.ClientConn, cluster raftpicker.RaftCluster) DispatcherServer {
+func NewRaftProxyDispatcherServer(local DispatcherServer, connSelector *raftpicker.ConnSelector, cluster raftpicker.RaftCluster) DispatcherServer {
 	return &raftProxyDispatcherServer{
-		local:   local,
-		cluster: cluster,
-		conn:    conn,
+		local:        local,
+		cluster:      cluster,
+		connSelector: connSelector,
 	}
 }
 
@@ -1137,13 +1137,21 @@ func (p *raftProxyDispatcherServer) Register(ctx context.Context, r *RegisterReq
 	md["redirect"] = append(md["redirect"], addr)
 	ctx = metadata.NewContext(ctx, md)
 
-	return NewDispatcherClient(p.conn).Register(ctx, r)
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return nil, err
+	}
+	return NewDispatcherClient(conn).Register(ctx, r)
 }
 
 func (p *raftProxyDispatcherServer) Session(r *SessionRequest, stream Dispatcher_SessionServer) error {
 
 	if p.cluster.IsLeader() {
 		return p.local.Session(r, stream)
+	}
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return err
 	}
 	var addr string
 	s, ok := transport.StreamFromContext(stream.Context())
@@ -1160,7 +1168,7 @@ func (p *raftProxyDispatcherServer) Session(r *SessionRequest, stream Dispatcher
 	md["redirect"] = append(md["redirect"], addr)
 	ctx := metadata.NewContext(stream.Context(), md)
 
-	clientStream, err := NewDispatcherClient(p.conn).Session(ctx, r)
+	clientStream, err := NewDispatcherClient(conn).Session(ctx, r)
 
 	if err != nil {
 		return err
@@ -1201,7 +1209,11 @@ func (p *raftProxyDispatcherServer) Heartbeat(ctx context.Context, r *HeartbeatR
 	md["redirect"] = append(md["redirect"], addr)
 	ctx = metadata.NewContext(ctx, md)
 
-	return NewDispatcherClient(p.conn).Heartbeat(ctx, r)
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return nil, err
+	}
+	return NewDispatcherClient(conn).Heartbeat(ctx, r)
 }
 
 func (p *raftProxyDispatcherServer) UpdateTaskStatus(ctx context.Context, r *UpdateTaskStatusRequest) (*UpdateTaskStatusResponse, error) {
@@ -1224,13 +1236,21 @@ func (p *raftProxyDispatcherServer) UpdateTaskStatus(ctx context.Context, r *Upd
 	md["redirect"] = append(md["redirect"], addr)
 	ctx = metadata.NewContext(ctx, md)
 
-	return NewDispatcherClient(p.conn).UpdateTaskStatus(ctx, r)
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return nil, err
+	}
+	return NewDispatcherClient(conn).UpdateTaskStatus(ctx, r)
 }
 
 func (p *raftProxyDispatcherServer) Tasks(r *TasksRequest, stream Dispatcher_TasksServer) error {
 
 	if p.cluster.IsLeader() {
 		return p.local.Tasks(r, stream)
+	}
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return err
 	}
 	var addr string
 	s, ok := transport.StreamFromContext(stream.Context())
@@ -1247,7 +1267,7 @@ func (p *raftProxyDispatcherServer) Tasks(r *TasksRequest, stream Dispatcher_Tas
 	md["redirect"] = append(md["redirect"], addr)
 	ctx := metadata.NewContext(stream.Context(), md)
 
-	clientStream, err := NewDispatcherClient(p.conn).Tasks(ctx, r)
+	clientStream, err := NewDispatcherClient(conn).Tasks(ctx, r)
 
 	if err != nil {
 		return err

--- a/manager/manager.go
+++ b/manager/manager.go
@@ -210,26 +210,12 @@ func (m *Manager) Run(ctx context.Context) error {
 		}
 	}()
 
-	api.RegisterCAServer(m.server, m.caserver)
-	api.RegisterRaftServer(m.server, m.raftNode)
-
-	errServe := make(chan error, 1)
-	go func() {
-		errServe <- m.server.Serve(lis)
-	}()
-
 	ctx = log.WithLogger(ctx, log.G(ctx).WithFields(
 		logrus.Fields{
 			"proto": lis.Addr().Network(),
 			"addr":  lis.Addr().String()}))
 	log.G(ctx).Info("listening")
 	go m.raftNode.Run(ctx)
-
-	leaderCtx, cancel := context.WithTimeout(ctx, 10*time.Second)
-	if err := raft.WaitForLeader(leaderCtx, m.raftNode); err != nil {
-		return err
-	}
-	cancel()
 
 	backoffConfig := *grpc.DefaultBackoffConfig
 	backoffConfig.MaxDelay = 2 * time.Second
@@ -239,19 +225,28 @@ func (m *Manager) Run(ctx context.Context) error {
 		grpc.WithTransportCredentials(m.config.SecurityConfig.ClientTLSCreds),
 	}
 
-	raftConn, err := raftpicker.Dial(m.raftNode, proxyOpts...)
-	if err != nil {
-		m.server.Stop()
-		return err
-	}
+	cs := raftpicker.NewConnSelector(m.raftNode, proxyOpts...)
 
 	localAPI := clusterapi.NewServer(m.raftNode.MemoryStore(), m.raftNode)
-	proxyAPI := api.NewRaftProxyClusterServer(localAPI, raftConn, m.raftNode)
+	proxyAPI := api.NewRaftProxyClusterServer(localAPI, cs, m.raftNode)
 
+	api.RegisterCAServer(m.server, m.caserver)
+	api.RegisterRaftServer(m.server, m.raftNode)
 	api.RegisterManagerServer(m.server, m)
 	api.RegisterClusterServer(m.server, proxyAPI)
 	api.RegisterDispatcherServer(m.server, m.dispatcher)
 
+	errServe := make(chan error, 1)
+	go func() {
+		errServe <- m.server.Serve(lis)
+	}()
+
+	leaderCtx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	if err := raft.WaitForLeader(leaderCtx, m.raftNode); err != nil {
+		m.server.Stop()
+		return err
+	}
+	cancel()
 	return <-errServe
 }
 

--- a/manager/raftpicker/raftpicker.go
+++ b/manager/raftpicker/raftpicker.go
@@ -8,8 +8,8 @@ import (
 	"google.golang.org/grpc/transport"
 )
 
-// Picker always picks address of cluster leader.
-type Picker struct {
+// picker always picks address of cluster leader.
+type picker struct {
 	mu   sync.Mutex
 	addr string
 	raft AddrSelector
@@ -17,20 +17,13 @@ type Picker struct {
 	cc   *grpc.ClientConn
 }
 
-// New returns new Picker with AddrSelector interface which it'll use for
-// picking. initAddr should be same as for Dial, because there is no way to get
-// target from ClientConn.
-func New(raft AddrSelector, initAddr string) grpc.Picker {
-	return &Picker{raft: raft, addr: initAddr}
-}
-
 // Init does initial processing for the Picker, e.g., initiate some connections.
-func (p *Picker) Init(cc *grpc.ClientConn) error {
+func (p *picker) Init(cc *grpc.ClientConn) error {
 	p.cc = cc
 	return nil
 }
 
-func (p *Picker) initConn() error {
+func (p *picker) initConn() error {
 	if p.conn == nil {
 		conn, err := grpc.NewConn(p.cc)
 		if err != nil {
@@ -43,7 +36,7 @@ func (p *Picker) initConn() error {
 
 // Pick blocks until either a transport.ClientTransport is ready for the upcoming RPC
 // or some error happens.
-func (p *Picker) Pick(ctx context.Context) (transport.ClientTransport, error) {
+func (p *picker) Pick(ctx context.Context) (transport.ClientTransport, error) {
 	p.mu.Lock()
 	if err := p.initConn(); err != nil {
 		p.mu.Unlock()
@@ -66,7 +59,7 @@ func (p *Picker) Pick(ctx context.Context) (transport.ClientTransport, error) {
 
 // PickAddr picks a peer address for connecting. This will be called repeated for
 // connecting/reconnecting.
-func (p *Picker) PickAddr() (string, error) {
+func (p *picker) PickAddr() (string, error) {
 	addr, err := p.raft.LeaderAddr()
 	if err != nil {
 		return "", err
@@ -78,35 +71,63 @@ func (p *Picker) PickAddr() (string, error) {
 }
 
 // State returns the connectivity state of the underlying connections.
-func (p *Picker) State() (grpc.ConnectivityState, error) {
+func (p *picker) State() (grpc.ConnectivityState, error) {
 	return p.conn.State(), nil
 }
 
 // WaitForStateChange blocks until the state changes to something other than
 // the sourceState. It returns the new state or error.
-func (p *Picker) WaitForStateChange(ctx context.Context, sourceState grpc.ConnectivityState) (grpc.ConnectivityState, error) {
+func (p *picker) WaitForStateChange(ctx context.Context, sourceState grpc.ConnectivityState) (grpc.ConnectivityState, error) {
 	return p.conn.WaitForStateChange(ctx, sourceState)
 }
 
 // Reset the current connection and force a reconnect to another address.
-func (p *Picker) Reset() error {
+func (p *picker) Reset() error {
 	p.conn.NotifyReset()
 	return nil
 }
 
 // Close closes all the Conn's owned by this Picker.
-func (p *Picker) Close() error {
+func (p *picker) Close() error {
 	return p.conn.Close()
 }
 
-// Dial returns *grpc.ClientConn with picker set to raftpicker with initial
-// address of cluster leader.
-func Dial(selector AddrSelector, opts ...grpc.DialOption) (*grpc.ClientConn, error) {
-	addr, err := selector.LeaderAddr()
+// ConnSelector is struct for obtaining connection with raftpicker.
+type ConnSelector struct {
+	mu      sync.Mutex
+	cc      *grpc.ClientConn
+	cluster RaftCluster
+	opts    []grpc.DialOption
+}
+
+// NewConnSelector returns new ConnSelector with cluster and grpc.DialOpts which
+// will be used for Dial on first call of Conn.
+func NewConnSelector(cluster RaftCluster, opts ...grpc.DialOption) *ConnSelector {
+	return &ConnSelector{
+		cluster: cluster,
+		opts:    opts,
+	}
+}
+
+// Conn returns *grpc.ClientConn with picker which picks raft cluster leader.
+// Internal connection estabilished lazily on this call.
+// It can return error if cluster wasn't ready at the moment of initial call.
+func (c *ConnSelector) Conn() (*grpc.ClientConn, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.cc != nil {
+		return c.cc, nil
+	}
+	addr, err := c.cluster.LeaderAddr()
 	if err != nil {
 		return nil, err
 	}
-	picker := New(selector, addr)
-	opts = append(opts, grpc.WithPicker(picker))
-	return grpc.Dial(addr, opts...)
+	picker := &picker{raft: c.cluster, addr: addr}
+	opts := append(c.opts, grpc.WithPicker(picker))
+	cc, err := grpc.Dial(addr, opts...)
+	if err != nil {
+		return nil, err
+	}
+	c.cc = cc
+	return c.cc, nil
 }

--- a/protobuf/plugin/raftproxy/raftproxy.go
+++ b/protobuf/plugin/raftproxy/raftproxy.go
@@ -27,17 +27,17 @@ func (g *raftProxyGen) Name() string {
 func (g *raftProxyGen) genProxyStruct(s *descriptor.ServiceDescriptorProto) {
 	g.gen.P("type " + serviceTypeName(s) + " struct {")
 	g.gen.P("\tlocal " + s.GetName() + "Server")
-	g.gen.P("\tconn *grpc.ClientConn")
+	g.gen.P("\tconnSelector *raftpicker.ConnSelector")
 	g.gen.P("\tcluster raftpicker.RaftCluster")
 	g.gen.P("}")
 }
 
 func (g *raftProxyGen) genProxyConstructor(s *descriptor.ServiceDescriptorProto) {
-	g.gen.P("func NewRaftProxy" + s.GetName() + "Server(local " + s.GetName() + "Server, conn *grpc.ClientConn, cluster raftpicker.RaftCluster) " + s.GetName() + "Server {")
+	g.gen.P("func NewRaftProxy" + s.GetName() + "Server(local " + s.GetName() + "Server, connSelector *raftpicker.ConnSelector, cluster raftpicker.RaftCluster) " + s.GetName() + "Server {")
 	g.gen.P("return &" + serviceTypeName(s) + `{
 		local: local,
 		cluster: cluster,
-		conn: conn,
+		connSelector: connSelector,
 	}`)
 	g.gen.P("}")
 }
@@ -87,9 +87,13 @@ func (g *raftProxyGen) genClientStreamingMethod(s *descriptor.ServiceDescriptorP
 	g.gen.P(`
 	if p.cluster.IsLeader() {
 			return p.local.` + m.GetName() + `(stream)
+	}
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return err
 	}`)
 	g.genStreamRedirectCheck()
-	g.gen.P("clientStream, err := New" + s.GetName() + "Client(p.conn)." + m.GetName() + "(ctx)")
+	g.gen.P("clientStream, err := New" + s.GetName() + "Client(conn)." + m.GetName() + "(ctx)")
 	g.gen.P(`
 	if err != nil {
 			return err
@@ -122,9 +126,13 @@ func (g *raftProxyGen) genServerStreamingMethod(s *descriptor.ServiceDescriptorP
 	g.gen.P(`
 	if p.cluster.IsLeader() {
 			return p.local.` + m.GetName() + `(r, stream)
+	}
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return err
 	}`)
 	g.genStreamRedirectCheck()
-	g.gen.P("clientStream, err := New" + s.GetName() + "Client(p.conn)." + m.GetName() + "(ctx, r)")
+	g.gen.P("clientStream, err := New" + s.GetName() + "Client(conn)." + m.GetName() + "(ctx, r)")
 	g.gen.P(`
 	if err != nil {
 			return err
@@ -151,9 +159,13 @@ func (g *raftProxyGen) genClientServerStreamingMethod(s *descriptor.ServiceDescr
 	g.gen.P(`
 	if p.cluster.IsLeader() {
 			return p.local.` + m.GetName() + `(stream)
+	}
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return err
 	}`)
 	g.genStreamRedirectCheck()
-	g.gen.P("clientStream, err := New" + s.GetName() + "Client(p.conn)." + m.GetName() + "(ctx)")
+	g.gen.P("clientStream, err := New" + s.GetName() + "Client(conn)." + m.GetName() + "(ctx)")
 	g.gen.P(`
 	if err != nil {
 			return err
@@ -208,9 +220,13 @@ func (g *raftProxyGen) genSimpleMethod(s *descriptor.ServiceDescriptorProto, m *
 	}
 	md["redirect"] = append(md["redirect"], addr)
 	ctx = metadata.NewContext(ctx, md)
-	`)
 
-	g.gen.P("return New" + s.GetName() + "Client(p.conn)." + m.GetName() + "(ctx, r)")
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return nil, err
+	}`)
+
+	g.gen.P("return New" + s.GetName() + "Client(conn)." + m.GetName() + "(ctx, r)")
 	g.gen.P("}")
 }
 

--- a/protobuf/plugin/raftproxy/test/raftproxy_test.go
+++ b/protobuf/plugin/raftproxy/test/raftproxy_test.go
@@ -54,10 +54,9 @@ func TestSimpleRedirect(t *testing.T) {
 	defer conn.Close()
 
 	cluster := &mockCluster{addr: addr}
-	cConn, err := raftpicker.Dial(cluster, grpc.WithInsecure(), grpc.WithTimeout(5*time.Second))
-	require.NoError(t, err)
+	cs := raftpicker.NewConnSelector(cluster, grpc.WithInsecure(), grpc.WithTimeout(5*time.Second))
 
-	api := NewRaftProxyRouteGuideServer(testRouteGuide{}, cConn, cluster)
+	api := NewRaftProxyRouteGuideServer(testRouteGuide{}, cs, cluster)
 	srv := grpc.NewServer()
 	RegisterRouteGuideServer(srv, api)
 	go srv.Serve(l)
@@ -78,10 +77,9 @@ func TestServerStreamRedirect(t *testing.T) {
 	defer conn.Close()
 
 	cluster := &mockCluster{addr: addr}
-	cConn, err := raftpicker.Dial(cluster, grpc.WithInsecure(), grpc.WithTimeout(5*time.Second))
-	require.NoError(t, err)
+	cs := raftpicker.NewConnSelector(cluster, grpc.WithInsecure(), grpc.WithTimeout(5*time.Second))
 
-	api := NewRaftProxyRouteGuideServer(testRouteGuide{}, cConn, cluster)
+	api := NewRaftProxyRouteGuideServer(testRouteGuide{}, cs, cluster)
 	srv := grpc.NewServer()
 	RegisterRouteGuideServer(srv, api)
 	go srv.Serve(l)
@@ -105,10 +103,9 @@ func TestClientStreamRedirect(t *testing.T) {
 	defer conn.Close()
 
 	cluster := &mockCluster{addr: addr}
-	cConn, err := raftpicker.Dial(cluster, grpc.WithInsecure(), grpc.WithTimeout(5*time.Second))
-	require.NoError(t, err)
+	cs := raftpicker.NewConnSelector(cluster, grpc.WithInsecure(), grpc.WithTimeout(5*time.Second))
 
-	api := NewRaftProxyRouteGuideServer(testRouteGuide{}, cConn, cluster)
+	api := NewRaftProxyRouteGuideServer(testRouteGuide{}, cs, cluster)
 	srv := grpc.NewServer()
 	RegisterRouteGuideServer(srv, api)
 	go srv.Serve(l)
@@ -134,10 +131,9 @@ func TestClientServerStreamRedirect(t *testing.T) {
 	defer conn.Close()
 
 	cluster := &mockCluster{addr: addr}
-	cConn, err := raftpicker.Dial(cluster, grpc.WithInsecure(), grpc.WithTimeout(5*time.Second))
-	require.NoError(t, err)
+	cs := raftpicker.NewConnSelector(cluster, grpc.WithInsecure(), grpc.WithTimeout(5*time.Second))
 
-	api := NewRaftProxyRouteGuideServer(testRouteGuide{}, cConn, cluster)
+	api := NewRaftProxyRouteGuideServer(testRouteGuide{}, cs, cluster)
 	srv := grpc.NewServer()
 	RegisterRouteGuideServer(srv, api)
 	go srv.Serve(l)

--- a/protobuf/plugin/raftproxy/test/service.pb.go
+++ b/protobuf/plugin/raftproxy/test/service.pb.go
@@ -713,16 +713,16 @@ func encodeVarintService(data []byte, offset int, v uint64) int {
 }
 
 type raftProxyRouteGuideServer struct {
-	local   RouteGuideServer
-	conn    *grpc.ClientConn
-	cluster raftpicker.RaftCluster
+	local        RouteGuideServer
+	connSelector *raftpicker.ConnSelector
+	cluster      raftpicker.RaftCluster
 }
 
-func NewRaftProxyRouteGuideServer(local RouteGuideServer, conn *grpc.ClientConn, cluster raftpicker.RaftCluster) RouteGuideServer {
+func NewRaftProxyRouteGuideServer(local RouteGuideServer, connSelector *raftpicker.ConnSelector, cluster raftpicker.RaftCluster) RouteGuideServer {
 	return &raftProxyRouteGuideServer{
-		local:   local,
-		cluster: cluster,
-		conn:    conn,
+		local:        local,
+		cluster:      cluster,
+		connSelector: connSelector,
 	}
 }
 
@@ -746,13 +746,21 @@ func (p *raftProxyRouteGuideServer) GetFeature(ctx context.Context, r *Point) (*
 	md["redirect"] = append(md["redirect"], addr)
 	ctx = metadata.NewContext(ctx, md)
 
-	return NewRouteGuideClient(p.conn).GetFeature(ctx, r)
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return nil, err
+	}
+	return NewRouteGuideClient(conn).GetFeature(ctx, r)
 }
 
 func (p *raftProxyRouteGuideServer) ListFeatures(r *Rectangle, stream RouteGuide_ListFeaturesServer) error {
 
 	if p.cluster.IsLeader() {
 		return p.local.ListFeatures(r, stream)
+	}
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return err
 	}
 	var addr string
 	s, ok := transport.StreamFromContext(stream.Context())
@@ -769,7 +777,7 @@ func (p *raftProxyRouteGuideServer) ListFeatures(r *Rectangle, stream RouteGuide
 	md["redirect"] = append(md["redirect"], addr)
 	ctx := metadata.NewContext(stream.Context(), md)
 
-	clientStream, err := NewRouteGuideClient(p.conn).ListFeatures(ctx, r)
+	clientStream, err := NewRouteGuideClient(conn).ListFeatures(ctx, r)
 
 	if err != nil {
 		return err
@@ -795,6 +803,10 @@ func (p *raftProxyRouteGuideServer) RecordRoute(stream RouteGuide_RecordRouteSer
 	if p.cluster.IsLeader() {
 		return p.local.RecordRoute(stream)
 	}
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return err
+	}
 	var addr string
 	s, ok := transport.StreamFromContext(stream.Context())
 	if ok {
@@ -810,7 +822,7 @@ func (p *raftProxyRouteGuideServer) RecordRoute(stream RouteGuide_RecordRouteSer
 	md["redirect"] = append(md["redirect"], addr)
 	ctx := metadata.NewContext(stream.Context(), md)
 
-	clientStream, err := NewRouteGuideClient(p.conn).RecordRoute(ctx)
+	clientStream, err := NewRouteGuideClient(conn).RecordRoute(ctx)
 
 	if err != nil {
 		return err
@@ -842,6 +854,10 @@ func (p *raftProxyRouteGuideServer) RouteChat(stream RouteGuide_RouteChatServer)
 	if p.cluster.IsLeader() {
 		return p.local.RouteChat(stream)
 	}
+	conn, err := p.connSelector.Conn()
+	if err != nil {
+		return err
+	}
 	var addr string
 	s, ok := transport.StreamFromContext(stream.Context())
 	if ok {
@@ -857,7 +873,7 @@ func (p *raftProxyRouteGuideServer) RouteChat(stream RouteGuide_RouteChatServer)
 	md["redirect"] = append(md["redirect"], addr)
 	ctx := metadata.NewContext(stream.Context(), md)
 
-	clientStream, err := NewRouteGuideClient(p.conn).RouteChat(ctx)
+	clientStream, err := NewRouteGuideClient(conn).RouteChat(ctx)
 
 	if err != nil {
 		return err

